### PR TITLE
prevent overriding shuffle settings in DataLoader for datapipes

### LINKED ISSUE #295: ShufflerIterDataPipe.set_shuffle_settings(True) gets overridden by DataLoader
Sorry if this isn't the right place for this bug report, please let me know.

I think there's an issue with the shuffling API. Doing

```py
dp = ShufflerIterDataPipe(dp)
dp.set_shuffle_settings(True)
```

won't actually shuffle anything unless `shuffle=True` is also passed to DataLoader.

This is because of

https://github.com/pytorch/pytorch/blob/3399876306772738936596f1711ec1b7f41495c8/torch/utils/data/dataloader.py#L231-L232 
and
https://github.com/pytorch/pytorch/blob/3399876306772738936596f1711ec1b7f41495c8/torch/utils/data/graph_settings.py#L30-L35 
and because of the fact that the default `shuffle` value of DataLoader is `False`.

### PR DESCRIPTION
**Pull request by pmeier**
Fixes https://github.com/pytorch/data/issues/295

Follow-up to https://github.com/pytorch/pytorch/pull/75014#issuecomment-1091921305. We only need to update locations where we actually check `shuffle` for identity with a boolean value, i.e. `shuffle is False`. For bool-ish checks like `if shuffle:`, `None` behaves just like `False`.

`IterDataPipe`'s are currently not mentioned in the docstring. Since this change only applies to them, I didn't update it. LMK, if I should do that.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -228,9 +228,9 @@ def __init__(self, dataset: Dataset[T_co], batch_size: Optional[int] = 1,
             # option. If this turns out to be useful in future, we can re-enable
             # this, and support custom samplers that specify the assignments to
             # specific workers.
-            if isinstance(dataset, IterDataPipe):
+            if isinstance(dataset, IterDataPipe) and shuffle is not None:
                 dataset = torch.utils.data.graph_settings.apply_shuffle_settings(dataset, shuffle=shuffle)
-            elif shuffle is not False:
+            elif shuffle:
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:233]: How about we convert `shuffle` back to `False` after the logic for `IterDataPipe` to preserve BC?

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -229,7 +229,8 @@ def __init__(self, dataset: Dataset[T_co], batch_size: Optional[int] = 1,
             # this, and support custom samplers that specify the assignments to
             # specific workers.
             if isinstance(dataset, IterDataPipe):
-                dataset = torch.utils.data.graph_settings.apply_shuffle_settings(dataset, shuffle=shuffle)
+                if shuffle is not None:
+                    dataset = torch.utils.data.graph_settings.apply_shuffle_settings(dataset, shuffle=shuffle)
             elif shuffle is not False:
                 raise ValueError(
                     "DataLoader with IterableDataset: expected unspecified "
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:236]: Could we check `elif shuffle is not False or shuffle is not None:` here to prevent BC breaking.

Comment (pmeier) [reply torch/utils/data/dataloader.py:236]: I thought you wanted 100% BC. With your suggestion passing `shuffle=None` will result in no error, but it currently does.

Comment (ejguan) [reply torch/utils/data/dataloader.py:236]: I think my proposed way is BC because we change the default value of `shuffle`. For users who use `IterableDataset` with default `shuffle` before, there won't be any Error. But, it currently does raise Error.

Comment (pmeier) [reply torch/utils/data/dataloader.py:236]: Take the following example:

```py
from torch.utils.data import DataLoader, IterableDataset


class Dataset(IterableDataset):
    def __iter__(self):
        return
        yield
```

- on `master`:

    ```py
    DataLoader(Dataset(), shuffle=False)  # passes
    DataLoader(Dataset(), shuffle=None)  # fails
    ```

- this PR

    ```py
    DataLoader(Dataset(), shuffle=False)  # passes
    DataLoader(Dataset(), shuffle=None)  # fails
    ```

- this PR with your suggestion

    ```py
    DataLoader(Dataset(), shuffle=False)  # fails
    DataLoader(Dataset(), shuffle=None)  # fails
    ```

Comment (ejguan) [reply torch/utils/data/dataloader.py:236]: I mean:
```py
DataLoader(Dataset())  # passes on master
DataLoader(Dataset())  # fails on this PR
```

Comment (pmeier) [reply torch/utils/data/dataloader.py:236]: That being said, there are failing tests related to this, that I need to take care of.

**Review by ejguan — APPROVED** [review]
Thank you for fixing it!
